### DISCUSSION TITLE
how can i update the pvc

### DISCUSSION THREAD

**Discussion by frank930211** [kubernetes-sigs/kubebuilder#3523] [Support] [answered]
eg:
        pvc, err := clientset.CoreV1().PersistentVolumeClaims("namespace").Get(context.TODO(), pvcName, metav1.GetOptions{})
	if err != nil {
		panic(err.Error())
	}
	pvc.Spec.Resources.Requests[corev1.ResourceStorage] = resource.MustParse(newSize)
	_, err = clientset.CoreV1().PersistentVolumeClaims("namespace").Update(context.TODO(), pvc, metav1.UpdateOptions{})
	if err != nil {
		panic(err.Error())
	}

i want to use reconciler.client to update but the error occurs:
"PersistentVolumeClaim \"myapp-pvc-myapp-0\" is invalid: spec: Forbidden: spec is immutable after creation except resources.requests for bound claims

it seems like that create a new pvc

**Comment by frank930211** [comment]
i dont want to use patch

**Comment by camilamacedo86** [comment, marked as answer]
The error you're seeing, "spec: Forbidden: spec is immutable after creation except resources.requests for bound claims", indicates that you're trying to modify a field of the PersistentVolumeClaim spec that isn't allowed to be changed after the PVC is created. In Kubernetes, once a PersistentVolumeClaim is created, most of its spec is immutable, with the exception of resources.requests for storage.

The code you've provided only modifies the resources.requests for storage, which should be allowed. However, if during your reconcile process, seems that other parts of the PVC spec are being inadvertently changed, you would get this error.

To update a PVC using the controller-runtime's client in a Kubebuilder-based reconciler, you can do the following:

- Fetch the existing PVC.
- Modify the resources.requests (or any other valid update).
- Update the PVC.

Here's how you can do that:

```go
// Fetch the existing PVC
pvc := &corev1.PersistentVolumeClaim{}
err := r.Client.Get(context.TODO(), types.NamespacedName{Name: pvcName, Namespace: "namespace"}, pvc)
if err != nil {
    // Handle error
}

// Modify the resources.requests
pvc.Spec.Resources.Requests[corev1.ResourceStorage] = resource.MustParse(newSize)

// Update the PVC
err = r.Client.Update(context.TODO(), pvc)
if err != nil {
    // Handle error
}
```

Closing this one as answered.
However, feel free to re-open if you need to do so.
